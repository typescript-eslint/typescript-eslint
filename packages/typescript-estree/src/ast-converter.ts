import { SourceFile } from 'typescript';
import { convertError, Converter, ASTMaps } from './convert';
import { convertComments } from './convert-comments';
import { convertTokens } from './node-utils';
import { Extra } from './parser-options';
import { TSESTree } from './ts-estree';
import { simpleTraverse } from './simple-traverse';

export function astConverter(
  ast: SourceFile,
  extra: Extra,
  shouldPreserveNodeMaps: boolean,
): { estree: TSESTree.Program; astMaps: ASTMaps | undefined } {
  /**
   * The TypeScript compiler produced fundamental parse errors when parsing the
   * source.
   */
  // internal typescript api...
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const parseDiagnostics = (ast as any).parseDiagnostics;
  if (parseDiagnostics.length) {
    throw convertError(parseDiagnostics[0]);
  }

  /**
   * Recursively convert the TypeScript AST into an ESTree-compatible AST
   */
  const instance = new Converter(ast, {
    errorOnUnknownASTType: extra.errorOnUnknownASTType || false,
    useJSXTextNode: extra.useJSXTextNode || false,
    shouldPreserveNodeMaps,
  });

  const estree = instance.convertProgram();

  /**
   * Optionally remove range and loc if specified
   */
  if (!extra.range || !extra.loc) {
    simpleTraverse(estree, {
      enter: node => {
        if (!extra.range) {
          delete node.range;
        }
        if (!extra.loc) {
          delete node.loc;
        }
      },
    });
  }

  /**
   * Optionally convert and include all tokens in the AST
   */
  if (extra.tokens) {
    estree.tokens = convertTokens(ast);
  }

  /**
   * Optionally convert and include all comments in the AST
   */
  if (extra.comment) {
    estree.comments = convertComments(ast, extra.code);
  }

  const astMaps = shouldPreserveNodeMaps ? instance.getASTMaps() : undefined;

  return { estree, astMaps };
}
