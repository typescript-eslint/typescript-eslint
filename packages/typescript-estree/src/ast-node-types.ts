/**
 * @fileoverview The AST node types produced by the parser.
 * @author Nicholas C. Zakas
 * @author James Henry <https://github.com/JamesHenry>
 * @copyright jQuery Foundation and other contributors, https://jquery.org/
 * MIT License
 */
export const AST_NODE_TYPES: { [key: string]: string } = {
  ArrayExpression: 'ArrayExpression',
  ArrayPattern: 'ArrayPattern',
  ArrowFunctionExpression: 'ArrowFunctionExpression',
  AssignmentExpression: 'AssignmentExpression',
  AssignmentPattern: 'AssignmentPattern',
  AwaitExpression: 'AwaitExpression',
  BigIntLiteral: 'BigIntLiteral',
  BinaryExpression: 'BinaryExpression',
  BlockStatement: 'BlockStatement',
  BreakStatement: 'BreakStatement',
  CallExpression: 'CallExpression',
  CatchClause: 'CatchClause',
  ClassBody: 'ClassBody',
  ClassDeclaration: 'ClassDeclaration',
  ClassExpression: 'ClassExpression',
  ClassImplements: 'ClassImplements',
  ClassProperty: 'ClassProperty',
  ConditionalExpression: 'ConditionalExpression',
  ContinueStatement: 'ContinueStatement',
  DebuggerStatement: 'DebuggerStatement',
  Decorator: 'Decorator',
  DoWhileStatement: 'DoWhileStatement',
  EmptyStatement: 'EmptyStatement',
  ExportAllDeclaration: 'ExportAllDeclaration',
  ExportDefaultDeclaration: 'ExportDefaultDeclaration',
  ExportNamedDeclaration: 'ExportNamedDeclaration',
  ExportSpecifier: 'ExportSpecifier',
  ExpressionStatement: 'ExpressionStatement',
  ForInStatement: 'ForInStatement',
  ForOfStatement: 'ForOfStatement',
  ForStatement: 'ForStatement',
  FunctionDeclaration: 'FunctionDeclaration',
  FunctionExpression: 'FunctionExpression',
  GenericTypeAnnotation: 'GenericTypeAnnotation',
  Identifier: 'Identifier',
  IfStatement: 'IfStatement',
  Import: 'Import',
  ImportDeclaration: 'ImportDeclaration',
  ImportDefaultSpecifier: 'ImportDefaultSpecifier',
  ImportNamespaceSpecifier: 'ImportNamespaceSpecifier',
  ImportSpecifier: 'ImportSpecifier',
  JSXAttribute: 'JSXAttribute',
  JSXClosingElement: 'JSXClosingElement',
  JSXClosingFragment: 'JSXClosingFragment',
  JSXElement: 'JSXElement',
  JSXEmptyExpression: 'JSXEmptyExpression',
  JSXExpressionContainer: 'JSXExpressionContainer',
  JSXFragment: 'JSXFragment',
  JSXIdentifier: 'JSXIdentifier',
  JSXMemberExpression: 'JSXMemberExpression',
  JSXNamespacedName: 'JSXNamespacedName',
  JSXOpeningElement: 'JSXOpeningElement',
  JSXOpeningFragment: 'JSXOpeningFragment',
  JSXSpreadAttribute: 'JSXSpreadAttribute',
  JSXSpreadChild: 'JSXSpreadChild',
  JSXText: 'JSXText',
  LabeledStatement: 'LabeledStatement',
  Literal: 'Literal',
  LogicalExpression: 'LogicalExpression',
  MemberExpression: 'MemberExpression',
  MetaProperty: 'MetaProperty',
  MethodDefinition: 'MethodDefinition',
  NewExpression: 'NewExpression',
  ObjectExpression: 'ObjectExpression',
  ObjectPattern: 'ObjectPattern',
  Program: 'Program',
  Property: 'Property',
  RestElement: 'RestElement',
  ReturnStatement: 'ReturnStatement',
  SequenceExpression: 'SequenceExpression',
  SpreadElement: 'SpreadElement',
  Super: 'Super',
  SwitchCase: 'SwitchCase',
  SwitchStatement: 'SwitchStatement',
  TaggedTemplateExpression: 'TaggedTemplateExpression',
  TemplateElement: 'TemplateElement',
  TemplateLiteral: 'TemplateLiteral',
  ThisExpression: 'ThisExpression',
  ThrowStatement: 'ThrowStatement',
  TryStatement: 'TryStatement',
  UnaryExpression: 'UnaryExpression',
  UpdateExpression: 'UpdateExpression',
  VariableDeclaration: 'VariableDeclaration',
  VariableDeclarator: 'VariableDeclarator',
  WhileStatement: 'WhileStatement',
  WithStatement: 'WithStatement',
  YieldExpression: 'YieldExpression',
  /**
   * TS-prefixed nodes
   */
  TSAbstractClassDeclaration: 'TSAbstractClassDeclaration',
  TSAbstractClassProperty: 'TSAbstractClassProperty',
  TSAbstractKeyword: 'TSAbstractKeyword',
  TSAbstractMethodDefinition: 'TSAbstractMethodDefinition',
  TSAnyKeyword: 'TSAnyKeyword',
  TSArrayType: 'TSArrayType',
  TSAsExpression: 'TSAsExpression',
  TSAsyncKeyword: 'TSAsyncKeyword',
  TSBooleanKeyword: 'TSBooleanKeyword',
  TSBigIntKeyword: 'TSBigIntKeyword',
  TSConditionalType: 'TSConditionalType',
  TSConstructorType: 'TSConstructorType',
  TSConstructSignature: 'TSConstructSignature',
  TSDeclareKeyword: 'TSDeclareKeyword',
  TSDeclareFunction: 'TSDeclareFunction',
  TSEnumDeclaration: 'TSEnumDeclaration',
  TSEnumMember: 'TSEnumMember',
  TSExportAssignment: 'TSExportAssignment',
  TSExportKeyword: 'TSExportKeyword',
  TSImportType: 'TSImportType',
  TSInferType: 'TSInferType',
  TSLiteralType: 'TSLiteralType',
  TSIndexedAccessType: 'TSIndexedAccessType',
  TSIndexSignature: 'TSIndexSignature',
  TSInterfaceBody: 'TSInterfaceBody',
  TSInterfaceDeclaration: 'TSInterfaceDeclaration',
  TSInterfaceHeritage: 'TSInterfaceHeritage',
  TSFunctionType: 'TSFunctionType',
  TSMethodSignature: 'TSMethodSignature',
  TSModuleBlock: 'TSModuleBlock',
  TSModuleDeclaration: 'TSModuleDeclaration',
  TSNamespaceFunctionDeclaration: 'TSNamespaceFunctionDeclaration',
  TSNonNullExpression: 'TSNonNullExpression',
  TSNeverKeyword: 'TSNeverKeyword',
  TSNullKeyword: 'TSNullKeyword',
  TSNumberKeyword: 'TSNumberKeyword',
  TSMappedType: 'TSMappedType',
  TSObjectKeyword: 'TSObjectKeyword',
  TSParameterProperty: 'TSParameterProperty',
  TSPrivateKeyword: 'TSPrivateKeyword',
  TSPropertySignature: 'TSPropertySignature',
  TSProtectedKeyword: 'TSProtectedKeyword',
  TSPublicKeyword: 'TSPublicKeyword',
  TSQualifiedName: 'TSQualifiedName',
  TSQuestionToken: 'TSQuestionToken',
  TSReadonlyKeyword: 'TSReadonlyKeyword',
  TSRestType: 'TSRestType',
  TSStaticKeyword: 'TSStaticKeyword',
  TSStringKeyword: 'TSStringKeyword',
  TSSymbolKeyword: 'TSSymbolKeyword',
  TSThisType: 'TSThisType',
  TSTypeAnnotation: 'TSTypeAnnotation',
  TSTypeAliasDeclaration: 'TSTypeAliasDeclaration',
  TSTypeLiteral: 'TSTypeLiteral',
  TSTypeOperator: 'TSTypeOperator',
  TSTypeParameter: 'TSTypeParameter',
  TSTypeParameterDeclaration: 'TSTypeParameterDeclaration',
  TSTypeParameterInstantiation: 'TSTypeParameterInstantiation',
  TSTypePredicate: 'TSTypePredicate',
  TSTypeReference: 'TSTypeReference',
  TSTypeQuery: 'TSTypeQuery',
  TSIntersectionType: 'TSIntersectionType',
  TSTupleType: 'TSTupleType',
  TSOptionalType: 'TSOptionalType',
  TSParenthesizedType: 'TSParenthesizedType',
  TSUnionType: 'TSUnionType',
  TSUndefinedKeyword: 'TSUndefinedKeyword',
  TSUnknownKeyword: 'TSUnknownKeyword',
  TSVoidKeyword: 'TSVoidKeyword'
};
