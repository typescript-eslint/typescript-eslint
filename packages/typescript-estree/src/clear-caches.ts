import { clearWatchCaches } from './create-program/getWatchProgramsForProjects';
import { clearProgramCache as clearProgramCacheOriginal } from './parser';
import { clearTSConfigMatchCache } from './parseSettings/createParseSettings';
import { clearGlobCache } from './parseSettings/resolveProjectList';

/**
 * Clears all of the internal caches.
 * Generally you shouldn't need or want to use this.
 * Examples of intended uses:
 * - In tests to reset parser state to keep tests isolated.
 * - In custom lint tooling that iteratively lints one project at a time to prevent OOMs.
 */
export function clearCaches(): void {
  clearProgramCacheOriginal();
  clearWatchCaches();
  clearTSConfigMatchCache();
  clearGlobCache();
}

// TODO - delete this in next major
export const clearProgramCache = clearCaches();
