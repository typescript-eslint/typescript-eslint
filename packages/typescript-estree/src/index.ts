export * from './parser';
export { ParserServices, TSESTreeOptions } from './parser-options';
export { simpleTraverse } from './simple-traverse';
export * from './ts-estree';
export { clearCaches } from './create-program/createWatchProgram';

// re-export for backwards-compat
export { visitorKeys } from '@typescript-eslint/visitor-keys';

// note - cannot migrate this to an import statement because it will make TSC copy the package.json to the dist folder
export const version: string = require('../package.json').version;
