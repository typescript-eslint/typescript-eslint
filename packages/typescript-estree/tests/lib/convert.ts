import { Converter } from '../../src/convert';
import ts from 'typescript';

describe('convert', () => {
  function convertCode(code: string): ts.SourceFile {
    return ts.createSourceFile(
      'text.ts',
      code,
      ts.ScriptTarget.ESNext,
      true,
      ts.ScriptKind.TSX
    );
  }

  it('deeplyCopy should convert node correctly', () => {
    const ast = convertCode('type foo = ?foo<T> | ?(() => void)?');

    const instance = new Converter(ast, {
      errorOnUnknownASTType: false,
      useJSXTextNode: false,
      shouldProvideParserServices: false
    });
    expect(instance.convertProgram()).toMatchSnapshot();
  });

  it('deeplyCopy should convert node with decorators correctly', () => {
    const ast = convertCode('@test class foo {}');

    const instance = new Converter(ast, {
      errorOnUnknownASTType: false,
      useJSXTextNode: false,
      shouldProvideParserServices: false
    });
    expect((instance as any).deeplyCopy(ast.statements[0])).toMatchSnapshot();
  });

  it('deeplyCopy should convert node with type parameters correctly', () => {
    const ast = convertCode('class foo<T> {}');

    const instance = new Converter(ast, {
      errorOnUnknownASTType: false,
      useJSXTextNode: false,
      shouldProvideParserServices: false
    });
    expect((instance as any).deeplyCopy(ast.statements[0])).toMatchSnapshot();
  });

  it('deeplyCopy should convert node with type arguments correctly', () => {
    const ast = convertCode('new foo<T>()');

    const instance = new Converter(ast, {
      errorOnUnknownASTType: false,
      useJSXTextNode: false,
      shouldProvideParserServices: false
    });
    expect(
      (instance as any).deeplyCopy((ast.statements[0] as any).expression)
    ).toMatchSnapshot();
  });

  it('deeplyCopy should convert array of nodes', () => {
    const ast = convertCode('new foo<T>()');

    const instance = new Converter(ast, {
      errorOnUnknownASTType: false,
      useJSXTextNode: false,
      shouldProvideParserServices: false
    });
    expect((instance as any).deeplyCopy(ast)).toMatchSnapshot();
  });

  it('deeplyCopy should fail on unknown node', () => {
    const ast = convertCode('type foo = ?foo<T> | ?(() => void)?');

    const instance = new Converter(ast, {
      errorOnUnknownASTType: true,
      useJSXTextNode: false,
      shouldProvideParserServices: false
    });
    expect(() => instance.convertProgram()).toThrow(
      'Unknown AST_NODE_TYPE: "TSJSDocNullableType"'
    );
  });

  it('nodeMaps should contain basic nodes', () => {
    const ast = convertCode(`
      'test';
      2;
      class foo {};
      type bar = {};
    `);

    const instance = new Converter(ast, {
      errorOnUnknownASTType: false,
      useJSXTextNode: false,
      shouldProvideParserServices: true
    });
    instance.convertProgram();
    const maps = instance.getASTMaps();

    function checkMaps(child: any) {
      child.forEachChild((node: any) => {
        if (
          node.kind !== ts.SyntaxKind.EndOfFileToken &&
          node.kind !== ts.SyntaxKind.JsxAttributes &&
          node.kind !== ts.SyntaxKind.VariableDeclaration
        ) {
          expect(node).toBe(
            maps.esTreeNodeToTSNodeMap.get(maps.tsNodeToESTreeNodeMap.get(node))
          );
        }
        checkMaps(node);
      });
    }

    expect(ast).toBe(
      maps.esTreeNodeToTSNodeMap.get(maps.tsNodeToESTreeNodeMap.get(ast))
    );
    checkMaps(ast);
  });

  it('nodeMaps should contain jsx nodes', () => {
    const ast = convertCode(`<a.b.c.d.e></a.b.c.d.e>`);

    const instance = new Converter(ast, {
      errorOnUnknownASTType: false,
      useJSXTextNode: false,
      shouldProvideParserServices: true
    });
    instance.convertProgram();
    const maps = instance.getASTMaps();

    function checkMaps(child: any) {
      child.forEachChild((node: any) => {
        if (
          node.kind !== ts.SyntaxKind.EndOfFileToken &&
          node.kind !== ts.SyntaxKind.JsxAttributes
        ) {
          expect(node).toBe(
            maps.esTreeNodeToTSNodeMap.get(maps.tsNodeToESTreeNodeMap.get(node))
          );
        }
        checkMaps(node);
      });
    }

    expect(ast).toBe(
      maps.esTreeNodeToTSNodeMap.get(maps.tsNodeToESTreeNodeMap.get(ast))
    );
    checkMaps(ast);
  });
});
