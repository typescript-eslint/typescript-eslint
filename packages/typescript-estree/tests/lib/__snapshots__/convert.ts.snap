// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert deeplyCopy should convert array of nodes 1`] = `
Object {
  "amdDependencies": Array [],
  "bindDiagnostics": Array [],
  "bindSuggestionDiagnostics": undefined,
  "checkJsDirective": undefined,
  "endOfFileToken": Object {
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
      },
      "start": Object {
        "column": 12,
        "line": 1,
      },
    },
    "range": Array [
      12,
      12,
    ],
    "type": "TSEndOfFileToken",
  },
  "externalModuleIndicator": undefined,
  "fileName": "text.ts",
  "hasNoDefaultLib": false,
  "identifierCount": 2,
  "identifiers": Map {
    "foo" => "foo",
    "T" => "T",
  },
  "isDeclarationFile": false,
  "languageVariant": 1,
  "languageVersion": 6,
  "libReferenceDirectives": Array [],
  "lineMap": Array [
    null,
  ],
  "loc": Object {
    "end": Object {
      "column": 12,
      "line": 1,
    },
    "start": Object {
      "column": 0,
      "line": 1,
    },
  },
  "nodeCount": 7,
  "parseDiagnostics": Array [],
  "pragmas": Map {},
  "range": Array [
    0,
    12,
  ],
  "referencedFiles": Array [],
  "scriptKind": 4,
  "statements": Array [
    Object {
      "expression": Object {
        "arguments": Array [],
        "callee": Object {
          "loc": Object {
            "end": Object {
              "column": 7,
              "line": 1,
            },
            "start": Object {
              "column": 4,
              "line": 1,
            },
          },
          "name": "foo",
          "range": Array [
            4,
            7,
          ],
          "type": "Identifier",
        },
        "loc": Object {
          "end": Object {
            "column": 12,
            "line": 1,
          },
          "start": Object {
            "column": 0,
            "line": 1,
          },
        },
        "range": Array [
          0,
          12,
        ],
        "type": "NewExpression",
        "typeParameters": Object {
          "loc": Object {
            "end": Object {
              "column": 10,
              "line": 1,
            },
            "start": Object {
              "column": 7,
              "line": 1,
            },
          },
          "params": Array [
            Object {
              "loc": Object {
                "end": Object {
                  "column": 9,
                  "line": 1,
                },
                "start": Object {
                  "column": 8,
                  "line": 1,
                },
              },
              "range": Array [
                8,
                9,
              ],
              "type": "TSTypeReference",
              "typeName": Object {
                "loc": Object {
                  "end": Object {
                    "column": 9,
                    "line": 1,
                  },
                  "start": Object {
                    "column": 8,
                    "line": 1,
                  },
                },
                "name": "T",
                "range": Array [
                  8,
                  9,
                ],
                "type": "Identifier",
              },
              "typeParameters": undefined,
            },
          ],
          "range": Array [
            7,
            10,
          ],
          "type": "TSTypeParameterInstantiation",
        },
      },
      "loc": Object {
        "end": Object {
          "column": 12,
          "line": 1,
        },
        "start": Object {
          "column": 0,
          "line": 1,
        },
      },
      "range": Array [
        0,
        12,
      ],
      "type": "ExpressionStatement",
    },
  ],
  "text": "new foo<T>()",
  "transformFlags": undefined,
  "type": "TSSourceFile",
  "typeReferenceDirectives": Array [],
}
`;

exports[`convert deeplyCopy should convert node correctly 1`] = `
Object {
  "body": Array [
    Object {
      "id": Object {
        "loc": Object {
          "end": Object {
            "column": 8,
            "line": 1,
          },
          "start": Object {
            "column": 5,
            "line": 1,
          },
        },
        "name": "foo",
        "range": Array [
          5,
          8,
        ],
        "type": "Identifier",
      },
      "loc": Object {
        "end": Object {
          "column": 35,
          "line": 1,
        },
        "start": Object {
          "column": 0,
          "line": 1,
        },
      },
      "range": Array [
        0,
        35,
      ],
      "type": "TSTypeAliasDeclaration",
      "typeAnnotation": Object {
        "loc": Object {
          "end": Object {
            "column": 35,
            "line": 1,
          },
          "start": Object {
            "column": 11,
            "line": 1,
          },
        },
        "range": Array [
          11,
          35,
        ],
        "transformFlags": undefined,
        "type": "TSJSDocNullableType",
        "typeAnnotation": Object {
          "loc": Object {
            "end": Object {
              "column": 35,
              "line": 1,
            },
            "start": Object {
              "column": 11,
              "line": 1,
            },
          },
          "range": Array [
            11,
            35,
          ],
          "type": "TSTypeAnnotation",
          "typeAnnotation": Object {
            "loc": Object {
              "end": Object {
                "column": 35,
                "line": 1,
              },
              "start": Object {
                "column": 12,
                "line": 1,
              },
            },
            "range": Array [
              12,
              35,
            ],
            "type": "TSUnionType",
            "types": Array [
              Object {
                "loc": Object {
                  "end": Object {
                    "column": 18,
                    "line": 1,
                  },
                  "start": Object {
                    "column": 12,
                    "line": 1,
                  },
                },
                "range": Array [
                  12,
                  18,
                ],
                "type": "TSTypeReference",
                "typeName": Object {
                  "loc": Object {
                    "end": Object {
                      "column": 15,
                      "line": 1,
                    },
                    "start": Object {
                      "column": 12,
                      "line": 1,
                    },
                  },
                  "name": "foo",
                  "range": Array [
                    12,
                    15,
                  ],
                  "type": "Identifier",
                },
                "typeParameters": Object {
                  "loc": Object {
                    "end": Object {
                      "column": 18,
                      "line": 1,
                    },
                    "start": Object {
                      "column": 15,
                      "line": 1,
                    },
                  },
                  "params": Array [
                    Object {
                      "loc": Object {
                        "end": Object {
                          "column": 17,
                          "line": 1,
                        },
                        "start": Object {
                          "column": 16,
                          "line": 1,
                        },
                      },
                      "range": Array [
                        16,
                        17,
                      ],
                      "type": "TSTypeReference",
                      "typeName": Object {
                        "loc": Object {
                          "end": Object {
                            "column": 17,
                            "line": 1,
                          },
                          "start": Object {
                            "column": 16,
                            "line": 1,
                          },
                        },
                        "name": "T",
                        "range": Array [
                          16,
                          17,
                        ],
                        "type": "Identifier",
                      },
                      "typeParameters": undefined,
                    },
                  ],
                  "range": Array [
                    15,
                    18,
                  ],
                  "type": "TSTypeParameterInstantiation",
                },
              },
              Object {
                "loc": Object {
                  "end": Object {
                    "column": 35,
                    "line": 1,
                  },
                  "start": Object {
                    "column": 21,
                    "line": 1,
                  },
                },
                "range": Array [
                  21,
                  35,
                ],
                "transformFlags": undefined,
                "type": "TSJSDocNullableType",
                "typeAnnotation": Object {
                  "loc": Object {
                    "end": Object {
                      "column": 35,
                      "line": 1,
                    },
                    "start": Object {
                      "column": 21,
                      "line": 1,
                    },
                  },
                  "range": Array [
                    21,
                    35,
                  ],
                  "type": "TSTypeAnnotation",
                  "typeAnnotation": Object {
                    "loc": Object {
                      "end": Object {
                        "column": 35,
                        "line": 1,
                      },
                      "start": Object {
                        "column": 22,
                        "line": 1,
                      },
                    },
                    "range": Array [
                      22,
                      35,
                    ],
                    "transformFlags": undefined,
                    "type": "TSJSDocNullableType",
                    "typeAnnotation": Object {
                      "loc": Object {
                        "end": Object {
                          "column": 34,
                          "line": 1,
                        },
                        "start": Object {
                          "column": 21,
                          "line": 1,
                        },
                      },
                      "range": Array [
                        21,
                        34,
                      ],
                      "type": "TSTypeAnnotation",
                      "typeAnnotation": Object {
                        "loc": Object {
                          "end": Object {
                            "column": 34,
                            "line": 1,
                          },
                          "start": Object {
                            "column": 22,
                            "line": 1,
                          },
                        },
                        "range": Array [
                          22,
                          34,
                        ],
                        "type": "TSParenthesizedType",
                        "typeAnnotation": Object {
                          "loc": Object {
                            "end": Object {
                              "column": 33,
                              "line": 1,
                            },
                            "start": Object {
                              "column": 23,
                              "line": 1,
                            },
                          },
                          "params": Array [],
                          "range": Array [
                            23,
                            33,
                          ],
                          "returnType": Object {
                            "loc": Object {
                              "end": Object {
                                "column": 33,
                                "line": 1,
                              },
                              "start": Object {
                                "column": 26,
                                "line": 1,
                              },
                            },
                            "range": Array [
                              26,
                              33,
                            ],
                            "type": "TSTypeAnnotation",
                            "typeAnnotation": Object {
                              "loc": Object {
                                "end": Object {
                                  "column": 33,
                                  "line": 1,
                                },
                                "start": Object {
                                  "column": 29,
                                  "line": 1,
                                },
                              },
                              "range": Array [
                                29,
                                33,
                              ],
                              "type": "TSVoidKeyword",
                            },
                          },
                          "type": "TSFunctionType",
                        },
                      },
                    },
                  },
                },
              },
            ],
          },
        },
      },
    },
  ],
  "loc": Object {
    "end": Object {
      "column": 35,
      "line": 1,
    },
    "start": Object {
      "column": 0,
      "line": 1,
    },
  },
  "range": Array [
    0,
    35,
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`convert deeplyCopy should convert node with decorators correctly 1`] = `
Object {
  "decorators": Array [
    Object {
      "expression": Object {
        "loc": Object {
          "end": Object {
            "column": 5,
            "line": 1,
          },
          "start": Object {
            "column": 1,
            "line": 1,
          },
        },
        "name": "test",
        "range": Array [
          1,
          5,
        ],
        "type": "Identifier",
      },
      "loc": Object {
        "end": Object {
          "column": 5,
          "line": 1,
        },
        "start": Object {
          "column": 0,
          "line": 1,
        },
      },
      "range": Array [
        0,
        5,
      ],
      "type": "Decorator",
    },
  ],
  "heritageClauses": undefined,
  "loc": Object {
    "end": Object {
      "column": 18,
      "line": 1,
    },
    "start": Object {
      "column": 0,
      "line": 1,
    },
  },
  "members": Array [],
  "modifiers": undefined,
  "name": Object {
    "loc": Object {
      "end": Object {
        "column": 15,
        "line": 1,
      },
      "start": Object {
        "column": 12,
        "line": 1,
      },
    },
    "name": "foo",
    "range": Array [
      12,
      15,
    ],
    "type": "Identifier",
  },
  "range": Array [
    0,
    18,
  ],
  "transformFlags": undefined,
  "type": "TSClassDeclaration",
  "typeParameters": null,
}
`;

exports[`convert deeplyCopy should convert node with type arguments correctly 1`] = `
Object {
  "arguments": Array [],
  "expression": Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 1,
      },
      "start": Object {
        "column": 4,
        "line": 1,
      },
    },
    "name": "foo",
    "range": Array [
      4,
      7,
    ],
    "type": "Identifier",
  },
  "loc": Object {
    "end": Object {
      "column": 12,
      "line": 1,
    },
    "start": Object {
      "column": 0,
      "line": 1,
    },
  },
  "range": Array [
    0,
    12,
  ],
  "transformFlags": undefined,
  "type": "TSNewExpression",
  "typeParameters": Object {
    "loc": Object {
      "end": Object {
        "column": 10,
        "line": 1,
      },
      "start": Object {
        "column": 7,
        "line": 1,
      },
    },
    "params": Array [
      Object {
        "loc": Object {
          "end": Object {
            "column": 9,
            "line": 1,
          },
          "start": Object {
            "column": 8,
            "line": 1,
          },
        },
        "range": Array [
          8,
          9,
        ],
        "type": "TSTypeReference",
        "typeName": Object {
          "loc": Object {
            "end": Object {
              "column": 9,
              "line": 1,
            },
            "start": Object {
              "column": 8,
              "line": 1,
            },
          },
          "name": "T",
          "range": Array [
            8,
            9,
          ],
          "type": "Identifier",
        },
        "typeParameters": undefined,
      },
    ],
    "range": Array [
      7,
      10,
    ],
    "type": "TSTypeParameterInstantiation",
  },
}
`;

exports[`convert deeplyCopy should convert node with type parameters correctly 1`] = `
Object {
  "heritageClauses": undefined,
  "loc": Object {
    "end": Object {
      "column": 15,
      "line": 1,
    },
    "start": Object {
      "column": 0,
      "line": 1,
    },
  },
  "members": Array [],
  "name": Object {
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 1,
      },
      "start": Object {
        "column": 6,
        "line": 1,
      },
    },
    "name": "foo",
    "range": Array [
      6,
      9,
    ],
    "type": "Identifier",
  },
  "range": Array [
    0,
    15,
  ],
  "transformFlags": undefined,
  "type": "TSClassDeclaration",
  "typeParameters": Object {
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
      },
      "start": Object {
        "column": 9,
        "line": 1,
      },
    },
    "params": Array [
      Object {
        "constraint": undefined,
        "default": undefined,
        "loc": Object {
          "end": Object {
            "column": 11,
            "line": 1,
          },
          "start": Object {
            "column": 10,
            "line": 1,
          },
        },
        "name": Object {
          "loc": Object {
            "end": Object {
              "column": 11,
              "line": 1,
            },
            "start": Object {
              "column": 10,
              "line": 1,
            },
          },
          "name": "T",
          "range": Array [
            10,
            11,
          ],
          "type": "Identifier",
        },
        "range": Array [
          10,
          11,
        ],
        "type": "TSTypeParameter",
      },
    ],
    "range": Array [
      9,
      12,
    ],
    "type": "TSTypeParameterDeclaration",
  },
}
`;
